src/main/java/org/apache/commons/cli/help/AbstractHelpFormatter.java

```diff
@@ -0,0 +1,327 @@
+/*
+  Licensed to the Apache Software Foundation (ASF) under one or more
+  contributor license agreements.  See the NOTICE file distributed with
+  this work for additional information regarding copyright ownership.
+  The ASF licenses this file to You under the Apache License, Version 2.0
+  (the "License"); you may not use this file except in compliance with
+  the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package org.apache.commons.cli.help;
+
+import static java.lang.String.format;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Function;
+
+import org.apache.commons.cli.Option;
+import org.apache.commons.cli.OptionGroup;
+import org.apache.commons.cli.Options;
+import org.apache.commons.cli.Util;
+
+/**
+ * The class for help formatters provides the framework to link the {@link HelpWriter} with the {@link OptionFormatter}
+ * and a default {@link TableDef} so to produce a standard format help page.
+ */
+public abstract class AbstractHelpFormatter {
```

Comment:
Let's not bake in an odd design for users because its uncomfortable for us developers to put in place a good one ;-) all for one special attribute. 

Extensible builders are normal and we do it in several places in Commons (Configuration, Exec, IO), the pattern is `[abstract or not] class Builder<T extends Builder<T>> implements Supplier<T>`  and `SubBuilder extends Builder<SubBuilder>`. 

A simple example for Commons Exec: 
- The parent builder https://github.com/apache/commons-exec/blob/e99527c1c7c42940bfda9ea8787db8c43573c4c3/src/main/java/org/apache/commons/exec/DefaultExecutor.java#L57
- The child builder: https://github.com/apache/commons-exec/blob/e99527c1c7c42940bfda9ea8787db8c43573c4c3/src/main/java/org/apache/commons/exec/DaemonExecutor.java#L34

Commons IO and Configuration have more advanced and complex examples.